### PR TITLE
Update to latest kubernetes version

### PR DESCRIPTION
**Pull request by surajssd**
_Note:_ Before merging, make sure that the bootkube PR is merged and then use the resulting merge commit ID here in `github.com/kinvolk/terraform-render-bootkube//?ref=`.

### CODE REVIEW

**Comment by surajssd** [comment]
This is working for Packet, AWS, KVM Libvirt !

Testing in progress Google Cloud and Azure!

**Comment by pothos** [comment]
@surajssd Please rebase and then change the `ci/…/` files to reflect the new versions.

**Comment by pothos** [comment]
Why is there a new option to set the cgroup driver while at the same time it is set by autodetecting the docker default (`/etc/kubernetes/configure-kubelet-cgroup-driver` writes to `/etc/kubernetes/kubelet.config`)? I think this new option has to be removed again. Did you add it as workaround for the bug in fixed in https://github.com/kinvolk/lokomotive-kubernetes/pull/62?

**Comment by pothos** [comment]
Fixed a merge bug where an old version was chosen for the libvirt provider.

File: aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:10]: Why this etcd version? I though 3.3.10 was tested by kubernetes upstream (see https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.15.md#unchanged). Am I missing something?

File: aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [inline aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Why is this needed to upgrade kubernetes version?

File: aws/flatcar-linux/kubernetes/controllers.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/controllers.tf:34]: Is this needed for the kubernetes upgrade? Or why is this added?

File: aws/flatcar-linux/kubernetes/controllers.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/controllers.tf]: Do we need to add support for flatcar edge in the same commit? Can this be done later, in a different PR? Or is it needed because of a bootkube dependency or minimize diff with typhoon?

File: aws/flatcar-linux/kubernetes/security.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/security.tf:54]: Uneeded diff? Seems the same is added at the end

File: aws/flatcar-linux/kubernetes/security.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/security.tf:65]: Why changing these ports?

File: azure/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [inline azure/flatcar-linux/kubernetes/bootkube.tf:9]: Why change all these variables in the upgrade? Is this needed to upgrade or is this a different *logical* change?

File: digital-ocean/container-linux/kubernetes/bootkube.tf
Comment (rata) [inline digital-ocean/container-linux/kubernetes/bootkube.tf:14]: Here we use another magic value than what is used for aws? Can you please elaborate how this was selected?

**Review by rata — COMMENTED** [review]
Thanks a lot for the patch!

I don't follow the reasoning on why so many changes are needed. Can you please explain?

I guess some are from following typhoon? And the rest is really needed for the upgrade or are changes that you made along the way but not related to the upgrade? I think it will be very useful for reviewing to clarify why the changes are done (either on the PR description, or different commits with a commit message that explains the reasonisn, etc.). 

IMHO, if possible, let's do the changes to upgrade kubernetes in one PR and the other changes, with their reasoning, etc. can be explained in another PR **if those are not related to the upgrade** . If these changes come from typhoon or something and makes sense to do them in the upgrade, fine, but in that case it would be great if there is any way (links to typhoon PRs?) to understand the changes done here.

Thanks again!
Rodrigo

File: aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (dongsupark) [reply aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:10]: If we want to address security issues, then I think we should also update etcd to v3.3.15, because it's compiled with Go 1.12.9.
For more details: https://github.com/etcd-io/etcd/blob/master/CHANGELOG-3.3.md#v3315-2019-08-19

**Comment by dongsupark** [comment]
I've just tested this PR on Packet with Flatcar stable. It worked fine.

**Comment by pothos** [comment]
FYI from the CI:
```
Getting source "github.com/kinvolk/terraform-render-bootkube?ref=f5e41c83ec01e00187cb5c50cb7e9ec8275bd43e"
Error downloading modules: Error loading modules: error downloading 'https://github.com/kinvolk/terraform-render-bootkube.git?ref=f5e41c83ec01e00187cb5c50cb7e9ec8275bd43e': /usr/bin/git exited with 128: fatal: reference is not a tree: f5e41c83ec01e00187cb5c50cb7e9ec8275bd43e
```

**Comment by surajssd** [comment]
> Why is there a new option to set the cgroup driver while at the same time it is set by autodetecting the docker default (/etc/kubernetes/configure-kubelet-cgroup-driver writes to /etc/kubernetes/kubelet.config)? I think this new option has to be removed again. Did you add it as workaround for the bug in fixed in #62?

@pothos I didn't notice your PR. The changes added in this PR are ported from upstream, do you suggest removal of that explicit checks? Removed those changes.

File: aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: This is not needed since we have a better way of dealing with this in an automated way #62.

File: aws/flatcar-linux/kubernetes/controllers.tf
Comment (surajssd) [reply aws/flatcar-linux/kubernetes/controllers.tf]: This is not needed since we have a better way of dealing with this in an automated way #62.

File: aws/flatcar-linux/kubernetes/controllers.tf
Comment (surajssd) [reply aws/flatcar-linux/kubernetes/controllers.tf:34]: Ported from:
https://github.com/poseidon/typhoon/commit/6db11d5908ff3e41a291e534598e6216af8eeb49

File: aws/flatcar-linux/kubernetes/security.tf
Comment (surajssd) [reply aws/flatcar-linux/kubernetes/security.tf:54]: Added from https://github.com/poseidon/typhoon/commit/af18296bc5ee8bf69745d236444df7a9d44c4050

File: aws/flatcar-linux/kubernetes/security.tf
Comment (surajssd) [reply aws/flatcar-linux/kubernetes/security.tf:65]: Added from https://github.com/poseidon/typhoon/commit/af18296bc5ee8bf69745d236444df7a9d44c4050

File: azure/flatcar-linux/kubernetes/bootkube.tf
Comment (surajssd) [reply azure/flatcar-linux/kubernetes/bootkube.tf:9]: This PR imports all the changes we missed from Typhoon.

File: digital-ocean/container-linux/kubernetes/bootkube.tf
Comment (surajssd) [reply digital-ocean/container-linux/kubernetes/bootkube.tf:14]: Added from https://github.com/poseidon/typhoon/commit/147c21a4bd69fb92e763f99ea4bb5738bdd1af65

File: azure/flatcar-linux/kubernetes/bootkube.tf
Comment (surajssd) [reply azure/flatcar-linux/kubernetes/bootkube.tf:9]: I can document those commits in this PR, do you know of a better of putting this information out there?

**Comment by pothos** [comment]
Maybe this PR should be called _import upstream changes_. Good if it helps us to sync better in the future. If there are formatting changes from squashed upstream changes, it's nice to have a separate commit but it's also more work to do so I wouldn't require it and rather have the process more often to reduce the change size.

**Comment by surajssd** [comment]
Upstream has moved to `0.12.*` of terraform and just rebasing isn't easy on that. I can call it import upstream no problem with it.

File: aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (rata) [reply aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:10]: Oh, cool. Thanks! These kind of info is important, imho, in the commit message

File: aws/flatcar-linux/kubernetes/security.tf
Comment (rata) [reply aws/flatcar-linux/kubernetes/security.tf:54]: Ohh, okay, thanks! This is really useful to review

File: azure/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [reply azure/flatcar-linux/kubernetes/bootkube.tf:9]: Doing that is certanly very useful! I didn't understand why there were so many changes :-D

I think doing that is fine now. When we switch to terraform 0.12, we can probably just merge. Then, the PR will be quite easy to review, it will merge all the commits and each commit has it's own explanation :)

File: digital-ocean/container-linux/kubernetes/bootkube.tf
Comment (rata) [reply digital-ocean/container-linux/kubernetes/bootkube.tf:14]: Thanks!

**Comment by rata** [comment]
@surajssd Ohh, I understand now. This PR is not only about upgrading kubernetes to latest, but also trying to backport changes from typhoon that is using another terraform version. Sorry I didn't notice before, now I see why there are so many changes here.

For the next time, I think we can either just update kubernetes (no sync with typhoon) or be very clear on which changes have been backported and probably have a 1:1 mapping (1 commit here for each commit upstream). That way, the backported commit can reference the original commit and is more easy to review.

Will try to review with this new scenario in mind now. Sorry for the noise :)

File: aws/flatcar-linux/kubernetes/security.tf
Comment (rata) [reply aws/flatcar-linux/kubernetes/security.tf:65]: I see, thanks for the pointer!

**Comment by surajssd** [comment]
Yep once we have moved to 0.12 we dont need to worry about where the changes are coming from and why!

**Review by rata — APPROVED** [review]
I've continued to look at this and tested this in Packet (with flatcar stable) and this LGTM!

**Comment by surajssd** [comment]
I will wait for one more approval and then we are good to go I think. Thanks @rata @dongsupark @pothos  for testing.

**Comment by rata** [comment]
@surajssd thank you!

@pothos Not sure if your comment was an approval and forgot to approve, or just a comment :)

**Comment by pothos** [comment]
Should `7e237ffa21fd85f76ddf2a215073aa7cd6ef2476` (the latest master commit) be used for bootkube?
Asking because I don't know if the referenced commit comes from the PR branch and if that was rebased on master before merging.

After that: Please merge ;) I've tested libvirt again.

**Comment by rata** [comment]
> Should `7e237ffa21fd85f76ddf2a215073aa7cd6ef2476` be used for bootkube?

Yes, exactly :)

**Comment by kosyfrances** [comment]
> Testing in progress Google Cloud and Azure!

Was this done :thinking:  
cc @surajssd

**Comment by rata** [comment]
@pothos if the bootkube commit is updated, can we merge this? :)

**Comment by surajssd** [comment]
@rata @pothos rebased and updated the PR also the tests pass! Can we merge this now?

**Review by pothos — APPROVED** [review]

**Pull request merged**
